### PR TITLE
Added inspection for possible plugin duplication

### PR DESCRIPTION
**Pull request by vasilii-b**
### Description (*)

This PR adds additional inspection to the plugin to check whenever there are duplicated plugin names in the project

image

### Fixed Issues (if relevant)

1. magento/magento2-phpstorm-plugin#48: Inspection for possible plugin duplication

### Questions or comments

N/A

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages

### CODE REVIEW

File: src/com/magento/idea/magento2plugin/stubs/indexes/PluginNameIndex.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    public KeyDescriptor<String> getKeyDescriptor() {
+        return myKeyDescriptor;
+    }
+
+    @NotNull
+    @Override
+    public FileBasedIndex.InputFilter getInputFilter() {
+        return file -> (
+                file.getFileType() == PhpFileType.INSTANCE
+                    || (file.getFileType() == XmlFileType.INSTANCE && file.getName().equals("di.xml"))
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/stubs/indexes/PluginNameIndex.java:121]: We have the constant for di.xml

File: src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+                    return null;
+                }
+                loadedFileHash.put(moduleXmlFilePath, virtualFile);
+                return virtualFile;
+            }
+
+            private String getModuleXmlFilePathByConfigFileDirectory(String fileDirectory, String fileArea) {
+                String moduleXmlFile = fileDirectory.replace(fileArea, "").concat(moduleXmlFileName);
+                if (fileDirectory.endsWith("etc")) {
+                    moduleXmlFile = fileDirectory.concat("/").concat(moduleXmlFileName);
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java:238]: File.separator

File: src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+                if (virtualFile == null) {
+                    return null;
+                }
+                loadedFileHash.put(moduleXmlFilePath, virtualFile);
+                return virtualFile;
+            }
+
+            private String getModuleXmlFilePathByConfigFileDirectory(String fileDirectory, String fileArea) {
+                String moduleXmlFile = fileDirectory.replace(fileArea, "").concat(moduleXmlFileName);
+                if (fileDirectory.endsWith("etc")) {
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java:237]: String literal

File: src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+                        continue;
+                    }
+
+                    String indexedPluginAttributeValue = ((XmlAttributeValue) indexedPlugin).getValue();
+                    if (!indexedPluginAttributeValue.equals(pluginNameAttributeValue)) {
+                        continue;
+                    }
+
+                    String indexedFileDirectory = indexedAttributeParent.getContainingDirectory().toString();
+                    String indexedFileFullPath = indexedFileDirectory.concat("/").concat(indexedAttributeParent.getName());
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java:141]: File.separator

File: src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+                            }
+                        }
+                    }
+                }
+            }
+
+            private List<HashMap<String, String>> fetchModuleNamesWhereSamePluginNameUsed(String pluginNameAttributeValue, String pluginTypeName, PluginIndex pluginIndex, PsiFile file) {
+                List<HashMap<String, String>> modulesName = new ArrayList<>();
+                String currentFileDirectory = file.getContainingDirectory().toString();
+                String currentFileFullPath = currentFileDirectory.concat("/").concat(file.getName());
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java:122]: File.separator

File: src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+                    String pluginNameAttributeValue = pluginNameAttribute.getValue();
+                    if (pluginNameAttributeValue == null) {
+                        continue;
+                    }
+
+                    List<XmlTag> targetObservers = fetchObserverTagsFromPluginTag(pluginXmlTag);
+
+                    for (XmlTag pluginTypeXmlTag: targetObservers) {
+                        XmlAttribute pluginTypeNameAttribute = pluginTypeXmlTag.getAttribute("name");
+                        XmlAttribute pluginTypeDisabledAttribute = pluginTypeXmlTag.getAttribute("disabled");
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java:77]: Add attributes and tags to file Class please

File: src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+
+                HashMap<String, XmlTag> targetObserversHash = new HashMap<>();
+
+                for (XmlTag pluginXmlTag: xmlTags) {
+                    HashMap<String, XmlTag> pluginProblems = new HashMap<>();
+                    if (!pluginXmlTag.getName().equals("type")) {
+                        continue;
+                    }
+
+                    XmlAttribute pluginNameAttribute = pluginXmlTag.getAttribute("name");
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java:66]: Attributes and tags should be in file Class.

File: src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+
+                if (xmlTags == null) {
+                    return;
+                }
+
+                HashMap<String, XmlTag> targetObserversHash = new HashMap<>();
+
+                for (XmlTag pluginXmlTag: xmlTags) {
+                    HashMap<String, XmlTag> pluginProblems = new HashMap<>();
+                    if (!pluginXmlTag.getName().equals("type")) {
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java:62]: Attributes and tags should be in file Class.

File: src/com/magento/idea/magento2plugin/stubs/indexes/PluginNameIndex.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        return methodReference.getName() != null && methodReference.getName().equals("dispatch");
+    }
+
+    private void grabPluginNamesFromXmlFile(XmlFile file, Map<String, Void> map) {
+        XmlDocument xmlDocument = file.getDocument();
+        if (xmlDocument != null) {
+            XmlTag xmlRootTag = xmlDocument.getRootTag();
+            if (xmlRootTag != null) {
+                for (XmlTag pluginTag : xmlRootTag.findSubTags("plugin")) {
+                    String name = pluginTag.getAttributeValue("name");
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/stubs/indexes/PluginNameIndex.java:101]: Attributes and tags should be in file Class.

**Comment by vasilii-b** [comment]
Hi @VitaliyBoyko,
Thank you very much for the feedback!
The requested changes were added.

**Comment by vasilii-b** [comment]
Hi @VitaliyBoyko,
As discussed, the redundant indexer was removed.
Can you please re-do the code-review?
Thank you!

File: src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+public class PluginDeclarationInspection extends PhpInspection {
+
+    @NotNull
+    @Override
+    public PsiElementVisitor buildVisitor(@NotNull ProblemsHolder problemsHolder, boolean b) {
+        return new XmlElementVisitor() {
+            private final String moduleXmlFileName = ModuleXml.getInstance().getFileName();
+            private static final String duplicatedPluginNameSameFileProblemDescription = "The plugin name already used in this file. For more details see Inspection Description.";
+            private static final String duplicatedPluginNameProblemDescription =
+                    "The plugin name \"%s\" for targeted \"%s\" class is already used in the module \"%s\" (%s scope). For more details see Inspection Description.";
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java:42]: We have begun using translation bundles for phrases like https://github.com/magento/magento2-phpstorm-plugin/blob/1.0.0-develop/resources/magento2/validation.properties. Could you please create the 'inspection' bundle and move all phrases there. Thank you!

Comment (VitaliyBoyko) [reply src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java:42]: @eduard13 really don’t like phrases within classes 😄

File: src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+import com.intellij.psi.xml.XmlDocument;
+import com.intellij.psi.xml.XmlTag;
+import com.jetbrains.php.lang.inspections.PhpInspection;
+import com.magento.idea.magento2plugin.indexes.PluginIndex;
+import com.magento.idea.magento2plugin.magento.files.ModuleDiXml;
+import com.magento.idea.magento2plugin.magento.files.ModuleXml;
+import com.magento.idea.magento2plugin.magento.packages.Package;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/inspections/xml/PluginDeclarationInspection.java:27]: And he also doesn't like empty lines like that :)

**Comment by vasilii-b** [comment]
Thank you so much @VitaliyBoyko  and @eduard13 for the translation bundle feature! 💪

**Comment by vasilii-b** [comment]
Thank you @VitaliyBoyko, will have a look.

**Pull request merged**